src/acl.c

```diff
@@ -2469,18 +2473,25 @@ void addACLLogEntry(client *c, int reason, int context, int argpos, sds username
     }
 }
 
-const char* getAclErrorMessage(int acl_res) {
-    /* Notice that a variant of this code also exists on aclCommand so
-     * it also need to be updated on changed. */
+sds getAclErrorMessage(int acl_res, user *user, struct redisCommand *cmd, sds errored_val, bool verbose) {
     switch (acl_res) {
     case ACL_DENIED_CMD:
-        return "can't run this command or subcommand";
+        return sdscatfmt(sdsempty(), "User %S has no permissions to run "
+                                     "the '%S' command", user->name, cmd->fullname);
     case ACL_DENIED_KEY:
-        return "can't access at least one of the keys mentioned in the command arguments";
+        if (verbose) {
+            return sdscatfmt(sdsempty(), "User %S has no permissions to access "
```

Comment:
note that key-names can contain null chars and newlines (user names and commands can't), so a verbose message is suitable for addReplyBulk and such, but not for addReplyError.
if we ever use it in such case, we need to use the trimming code in addReplyErrorFormatInternal

Reply:
is it a better idea that it should be returned in a trimmed form at this point?

Reply:
maybe it's a good idea... chances are these errors are gonna be used as errors by the caller in some way.
currently the only caller that uses verbose is ACL DRYRUN, right?
@madolson WDTY?